src/Features/LanguageServer/CommonLanguageServerProtocol.Framework/IRequestDispatcher.cs

```diff
@@ -0,0 +1,33 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+
+#nullable enable
+
+namespace CommonLanguageServerProtocol.Framework;
+
+public interface IRequestDispatcher<RequestContextType>
+{
+    ImmutableArray<RequestHandlerMetadata> GetRegisteredMethods();
+
+    Task<TResponseType?> ExecuteRequestAsync<TRequestType, TResponseType>(
+        string methodName,
+        TRequestType request,
+        IRequestExecutionQueue<RequestContextType> queue,
```

Comment:
i'm a little unclear on why the queue is passed along to these requests.  what's the idea there?  (in other words, i would have guessed normally that hte impl of the dispatcher just owns a queue it uses).  THis feels odd, to me, as a parameter as it implies that someone could pass along different queues for different requests?  does that happen?

Reply:
This is an artifact of [the Roslyn implementation I based things on](https://github.com/dotnet/roslyn/blob/main/src/Features/LanguageServer/Protocol/RequestDispatcher.cs#L85). I think you're correct though that this abstraction makes the ownership and control flow unclear. I think it might be more appropriate if instead of `IRequestDispatcher` this was more like a `IHandlerProvider` which was called internally by `IRequestExecutionQueue` to be given the handler object, and `IRequestExecutionQueue` took over responsibilities for "dispatching" to itself. Hopefully that makes sense.